gobblin-runtime/src/main/java/gobblin/runtime/Task.java

```diff
@@ -233,6 +244,52 @@ public void run() {
     }
   }
 
+  private void failTask(Throwable t) {
+    LOG.error(String.format("Task %s failed", this.taskId), t);
+    this.taskState.setWorkingState(WorkUnitState.WorkingState.FAILED);
+    this.taskState.setProp(ConfigurationKeys.TASK_FAILURE_EXCEPTION_KEY, ExceptionUtils.getStackTrace(t));
+  }
+
+  /**
+   * Whether the task should directly publish data to the final publisher output dir.
+   * The task should publish data if {@link ConfigurationKeys#JOB_COMMIT_POLICY_KEY} is set to "partial", and
+   * {@link ConfigurationKeys#PUBLISH_DATA_IN_TASKS} is set to true, and task state is SUCCESSFUL.
+   */
+  private boolean shouldPublishDataInTask() {
+    boolean jobCommitPolicyIsPartial =
+        JobCommitPolicy.getCommitPolicy(this.taskState.getProperties()) == JobCommitPolicy.COMMIT_ON_PARTIAL_SUCCESS;
+    boolean publishDataInTasksAllowed = this.taskState.getPropAsBoolean(ConfigurationKeys.PUBLISH_DATA_IN_TASKS,
+        ConfigurationKeys.DEFAULT_PUBLISH_DATA_IN_TASKS);
+
+    if (jobCommitPolicyIsPartial && !publishDataInTasksAllowed) {
+      LOG.info(String.format("%s is partial, but %s is disabled. Will publish data in driver.",
+          ConfigurationKeys.JOB_COMMIT_POLICY_KEY, ConfigurationKeys.PUBLISH_DATA_IN_TASKS));
+    } else if (!jobCommitPolicyIsPartial && publishDataInTasksAllowed) {
+      LOG.info(String.format("%s is enabled, but %s is set to full. Will publish data in driver.",
+          ConfigurationKeys.PUBLISH_DATA_IN_TASKS, ConfigurationKeys.JOB_COMMIT_POLICY_KEY));
+    }
+
+    return jobCommitPolicyIsPartial && publishDataInTasksAllowed
+        && this.taskState.getWorkingState() == WorkUnitState.WorkingState.SUCCESSFUL;
```

Comment:
Actually currently if partial commit is selected, failed tasks are also allowed to publish data.